clang-tools-extra/clang-tidy/portability/TemplateVirtualMemberFunctionCheck.cpp

```diff
@@ -0,0 +1,48 @@
+//===--- TemplateVirtualMemberFunctionCheck.cpp - clang-tidy --------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "TemplateVirtualMemberFunctionCheck.h"
+#include "clang/ASTMatchers/ASTMatchFinder.h"
+
+using namespace clang::ast_matchers;
+
+namespace clang::tidy::portability {
+
+void TemplateVirtualMemberFunctionCheck::registerMatchers(MatchFinder *Finder) {
+  Finder->addMatcher(classTemplateSpecializationDecl().bind("specialization"),
+                     this);
+}
+
+void TemplateVirtualMemberFunctionCheck::check(
+    const MatchFinder::MatchResult &Result) {
+  const auto *MatchedDecl =
+      Result.Nodes.getNodeAs<ClassTemplateSpecializationDecl>("specialization");
+
+  if (MatchedDecl->isExplicitSpecialization())
+    return;
```

Comment:
I think you would only want to ignore full specializations, partial specializations are still templates. I.e.: `MatchedDecl->isExplicitSpecialization() && !llvm::isa<ClassTemplatePartialSpecializationDecl>(MatchedDecl)`

Reply:
I think we can ignore partial specializations too. A partial specialization is still a template, except it has _some_ parameters specified. Because of it still being a template, the same rules should apply for instantiating its member functions.

For explicit specializations, the standard explicitly specifies that it's member functions should be treated as methods of regular classes and I think that's why they are instantiated even if they are not used.

> § 13.9.4 6 [...] Members of an explicitly specialized class template are defined in
the same manner as members of normal classes, and not using the template<> syntax. [...]

Reply:
You're right

Reply:
Actually, see https://godbolt.org/z/Wv91aaTcx

```c++
template <typename T, typename U = T>
struct CrossPlatformError {
    virtual ~CrossPlatformError() = default;

    static void used() {}

    virtual void unused() { T MSVCError = this; };
};
template <typename U>
struct CrossPlatformError<int, U> {
    virtual ~CrossPlatformError() = default;
    static void used() {}
    virtual void different_unused() { U MSVCError = this; };
};

int main() {
    CrossPlatformError<int>::used();
    return 0;
}
```

partial specializations could introduce the same issue, and only fully specialized class templates should be ignored.

Reply:
In this case there is an implicit instantiation, which results in the `ClassTemplateSpecializationDecl <line:9:1, line:14:1> line:10:8 struct CrossPlatformError definition implicit_instantiation` node. If you delete the `CrossPlatformError<int>::used();` line, the error is not reported even though the partial specialization is present.

This case is detected by the current implementation, I also added a testcase about it recently.

Reply:
My bad.
I somehow got hung up on the partial specialization itself, but it just adds the possibility of the issue. The fact that the member is unused is only really knowable once we have an instantiation, which triggers a specialization for that instantiation.